wit/atomic.wit

```diff
@@ -1,31 +1,32 @@
 /// A keyvalue interface that provides atomic operations.
 /// 
-/// Atomic operations are single, indivisible operations. When a fault causes
-/// an atomic operation to fail, it will appear to the invoker of the atomic
-/// operation that the action either completed successfully or did nothing
-/// at all.
-interface atomic {
-	/// A keyvalue interface that provides atomic operations.
-	use types.{bucket, error, key};
+/// Atomic operations are single, indivisible operations. When a fault causes an atomic operation to
+/// fail, it will appear to the invoker of the atomic operation that the action either completed
+/// successfully or did nothing at all.
+interface atomics {
+  use types.{bucket, error};
+  
+  /// The atomics resource contains atomic operations that can be executed against a store
+  resource atomics {
+	/// Create a new atomic resource, using the given bucket to access the data
+	open: static func(store: borrow<bucket>) -> atomics;
 
-	/// Atomically increment the value associated with the key in the bucket by the 
-	/// given delta. It returns the new value.
+	/// Atomically increment the value associated with the key in the store by the given delta. It
+	/// returns the new value.
 	///
-	/// If the key does not exist in the bucket, it creates a new key-value pair
-	/// with the value set to the given delta. 
+	/// If the key does not exist in the store, it creates a new key-value pair with the value set
+	/// to the given delta. 
 	///
 	/// If any other error occurs, it returns an `Err(error)`.
-	increment: func(bucket: borrow<bucket>, key: key, delta: u64) -> result<u64, error>;
+	increment: func(key: string, delta: u64) -> result<u64, error>;
 	
-	/// Compare-and-swap (CAS) atomically updates the value associated with the key
-	/// in the bucket if the value matches the old value. This operation returns
-	/// `Ok(true)` if the swap was successful, `Ok(false)` if the value did not match,
-	/// 
-	/// A successful CAS operation means the current value matched the `old` value
-	/// and was replaced with the `new` value.
+	/// Compare-and-swap (CAS) atomically updates the value associated with the key in the store if
+	/// the `cas-id` matches the current `cas-id` in the store. This operation returns `Ok(true)` if
+	/// the swap was successful, `Ok(false)` if the `cas-id` did not match.
 	///
-	/// If the key does not exist in the bucket, it returns `Ok(false)`.
+	/// If the key does not exist in the store, it returns `Ok(false)`.
 	/// 
 	/// If any other error occurs, it returns an `Err(error)`.
-	compare-and-swap: func(bucket: borrow<bucket>, key: key, old: u64, new: u64) -> result<bool, error>;
+	compare-and-swap: func(key: string, value: list<u8>, cas-id: u64) -> result<bool, error>;
```

Comment:
Please note: I did some digging here and for things that _do_ support CAS operations, it seemed like they all mostly used a numeric ID that was used to do the actual compare step. If someone else has additional ideas or guidance here, please let me know. We can also remove this functionality if we can't find a relatively standard way to do this

Reply:
This was actually an open question I had and forgot to put in my comment. I don't quite know what to do with it. My first, somewhat naive approach, was to reserve `0` as the number for a value that hasn't been set yet. Completely open to ideas here.

Based on my investigations, I think that everything had an ID that would automatically change/increment the ID, even if CAS wasn't used to set the value. Not _super_ confident in that, but that is what I went with.

This also reminded me about why this might not end up working very well. In order to use an ID to compare and swap, all the implementations I've used have a mechanism for fetching the current ID when you fetch the value. We don't have any concept of that in the current interface, and we'd have to add it everywhere, even if people aren't using atomics. It might be worth maybe deferring the addition of this function to a later version

Reply:
Yeah, perhaps a good idea to take it out for the moment.  These are pretty subtle and impactful design choices, so if there's not an "obvious" right precedent here, perhaps we can pick it up in the next iteration after some focused investigation.

Reply:
Ok, that sounds best to me. Will remove